File: qball-solver/solver.js
/* MFT Q-Ball Lepton Spectrum Solver — JavaScript layer (v2)
 *
 * Scan ω² internally, return the full discrete tower, highlight the
 * canonical lepton triple. ω² is never an input.
 */

let pyodide = null;
let solverReady = false;
let activeSector = 'lepton_sector';  // tracks which preset is active

const statusText = document.getElementById('status-text');
const progressFill = document.getElementById('progress-fill');
const statusSection = document.getElementById('status');
const solverUI = document.getElementById('solver-ui');

function setProgress(pct, msg) {
    progressFill.style.width = `${pct}%`;
    if (msg) statusText.textContent = msg;
}

async function initPyodide() {
    try {
        setProgress(10, 'Loading Pyodide runtime…');
        pyodide = await loadPyodide();

        setProgress(40, 'Loading NumPy and SciPy…');
        await pyodide.loadPackage(['numpy', 'scipy']);

        setProgress(80, 'Loading MFT solver…');
        const solverPyResp = await fetch('solver.py');
        const solverPyText = await solverPyResp.text();
        pyodide.runPython(solverPyText);

        setProgress(100, 'Ready.');
        solverReady = true;

        setTimeout(() => {
            statusSection.style.display = 'none';
            solverUI.style.display = 'grid';
        }, 400);
    } catch (err) {
        statusText.textContent = `Error loading solver: ${err.message}`;
        progressFill.style.background = '#c0392b';
        console.error(err);
    }
}

// === Plotting helpers ===

function clearCanvas(canvas) {
    const ctx = canvas.getContext('2d');
    ctx.clearRect(0, 0, canvas.width, canvas.height);
}

function plotCurves(canvas, curves, options = {}) {
    /* curves: array of { xs, ys, color, label } */
    const ctx = canvas.getContext('2d');
    const W = canvas.width;
    const H = canvas.height;

    const padding = { left: 50, right: 100, top: 20, bottom: 35 };
    const plotW = W - padding.left - padding.right;
    const plotH = H - padding.top - padding.bottom;

    clearCanvas(canvas);

    if (curves.length === 0) {
        ctx.fillStyle = '#888';
        ctx.font = '13px sans-serif';
        ctx.textAlign = 'center';
        ctx.fillText('No data', W / 2, H / 2);
        return;
    }

    // Auto-range across all curves
    let xmin = Infinity, xmax = -Infinity, ymin = Infinity, ymax = -Infinity;
    for (const c of curves) {
        for (const x of c.xs) { if (x < xmin) xmin = x; if (x > xmax) xmax = x; }
        for (const y of c.ys) { if (y < ymin) ymin = y; if (y > ymax) ymax = y; }
    }
    if (options.xmin !== undefined) xmin = options.xmin;
    if (options.xmax !== undefined) xmax = options.xmax;
    if (options.ymin !== undefined) ymin = options.ymin;
    if (options.ymax !== undefined) ymax = options.ymax;
    const yrange = ymax - ymin || 1;
    const xrange = xmax - xmin || 1;

    const xToPixel = x => padding.left + ((x - xmin) / xrange) * plotW;
    const yToPixel = y => padding.top + plotH - ((y - ymin) / yrange) * plotH;

    // Background
    ctx.fillStyle = '#fafafa';
    ctx.fillRect(padding.left, padding.top, plotW, plotH);

    // Axes
    ctx.strokeStyle = '#333';
    ctx.lineWidth = 1;
    ctx.beginPath();
    ctx.moveTo(padding.left, padding.top);
    ctx.lineTo(padding.left, padding.top + plotH);
    ctx.lineTo(padding.left + plotW, padding.top + plotH);
    ctx.stroke();

    // Zero line
    if (ymin < 0 && ymax > 0) {
        const yZero = yToPixel(0);
        ctx.strokeStyle = '#aaa';
        ctx.setLineDash([3, 3]);
        ctx.beginPath();
        ctx.moveTo(padding.left, yZero);
        ctx.lineTo(padding.left + plotW, yZero);
        ctx.stroke();
        ctx.setLineDash([]);
    }

    // Tick labels
    ctx.fillStyle = '#444';
    ctx.font = '11px sans-serif';
    ctx.textAlign = 'right';
    ctx.fillText(ymax.toFixed(2), padding.left - 5, padding.top + 4);
    ctx.fillText(ymin.toFixed(2), padding.left - 5, padding.top + plotH);
    ctx.textAlign = 'center';
    ctx.fillText(xmin.toFixed(1), padding.left, padding.top + plotH + 14);
    ctx.fillText(xmax.toFixed(1), padding.left + plotW, padding.top + plotH + 14);

    if (options.xlabel) {
        ctx.textAlign = 'center';
        ctx.font = '12px sans-serif';
        ctx.fillText(options.xlabel, padding.left + plotW / 2, H - 8);
    }
    if (options.ylabel) {
        ctx.save();
        ctx.translate(15, padding.top + plotH / 2);
        ctx.rotate(-Math.PI / 2);
        ctx.textAlign = 'center';
        ctx.fillText(options.ylabel, 0, 0);
        ctx.restore();
    }

    // Curves
    for (const c of curves) {
        ctx.strokeStyle = c.color || '#2c5aa0';
        ctx.lineWidth = c.lineWidth || 2;
        ctx.beginPath();
        for (let i = 0; i < c.xs.length; i++) {
            const px = xToPixel(c.xs[i]);
            const py = yToPixel(c.ys[i]);
            if (i === 0) ctx.moveTo(px, py);
            else ctx.lineTo(px, py);
        }
        ctx.stroke();
    }

    // Markers
    if (options.markers) {
        for (const m of options.markers) {
            const px = xToPixel(m.x);
            ctx.strokeStyle = m.color || '#888';
            ctx.lineWidth = 1.5;
            ctx.setLineDash([4, 3]);
            ctx.beginPath();
            ctx.moveTo(px, padding.top);
            ctx.lineTo(px, padding.top + plotH);
            ctx.stroke();
            ctx.setLineDash([]);
            if (m.label) {
                ctx.fillStyle = m.color || '#888';
                ctx.font = '10px sans-serif';
                ctx.textAlign = 'left';
                ctx.fillText(m.label, px + 3, padding.top + 12);
            }
        }
    }

    // Legend
    if (curves.some(c => c.label)) {
        let yLegend = padding.top + 10;
        const xLegend = padding.left + plotW + 10;
        ctx.font = '11px sans-serif';
        ctx.textAlign = 'left';
        for (const c of curves) {
            if (!c.label) continue;
            ctx.strokeStyle = c.color;
            ctx.lineWidth = 2;
            ctx.beginPath();
            ctx.moveTo(xLegend, yLegend);
            ctx.lineTo(xLegend + 18, yLegend);
            ctx.stroke();
            ctx.fillStyle = '#333';
            ctx.fillText(c.label, xLegend + 22, yLegend + 4);
            yLegend += 16;
        }
    }
}

// === Render results ===

const LEPTON_COLORS = {
    electron: '#27ae60',
    muon: '#2c5aa0',
    tau: '#c0392b',
};

const LEPTON_OBSERVED_MEV = {
    electron: 0.511,
    muon: 105.66,
    tau: 1776.86,
};

function renderResults(result) {
    const summary = document.getElementById('results-summary');
    const tripleContainer = document.getElementById('lepton-triple-container');
    const tableContainer = document.getElementById('spectrum-table-container');

    if (!result.success) {
        summary.innerHTML = `<p class="error">${result.message}</p>`;
        clearCanvas(document.getElementById('potential-canvas'));
        clearCanvas(document.getElementById('profile-canvas'));
        tableContainer.innerHTML = '';
        tripleContainer.innerHTML = '';
        return;
    }

    // === Summary ===
    const sr = result.silver_ratio;
    const srBadge = sr.satisfied
        ? '<span class="badge ok">silver-ratio satisfied</span>'
        : `<span class="badge warn">silver-ratio violated (λ₄²=${sr.lam4_squared.toFixed(3)}, 8m₂λ₆=${sr.eight_m2_lam6.toFixed(3)})</span>`;

    const sectorIsLepton = result.sector === 'lepton_sector';
    summary.innerHTML = `
        <p>
            <strong>${result.spectrum.length}</strong> distinct solitons found in the spectrum.
            ${srBadge}
        </p>
        <p class="hint">
            Mass column uses MFT_TO_MEV = ${result.mft_to_mev.toFixed(2)},
            the global conversion factor anchored by the lepton ground-state at the canonical potential
            (m_e = 0.511 MeV). ${sectorIsLepton
                ? 'For the lepton sector this is the correct calibration.'
                : 'Non-lepton sectors apply the same conversion; matching to observed masses in those sectors may require additional MFT-physics adjustments (sector-specific calibration anchors, mass-extraction formulas, or back-reaction corrections) not yet implemented here.'
            }
            φ_barrier = ${result.phi_barrier.toFixed(4)}, φ_vacuum = ${result.phi_vacuum.toFixed(4)}.
        </p>
    `;

    // === Lepton triple ===
    // Only show the lepton-triple table when the user has loaded the lepton
    // sector preset; for quark/boson sectors, even if triple-matching ran,
    // the labels (electron/muon/tau) wouldn't make sense.
    const triple = (result.sector === 'lepton_sector') ? result.lepton_triple : null;
    if (triple) {
        const e = result.spectrum[triple.electron_idx];
        const mu = result.spectrum[triple.muon_idx];
        const ta = result.spectrum[triple.tau_idx];

        const errorPct = (model, observed) =>
            (100 * Math.abs(model - observed) / observed).toFixed(1);

        const r10_model = mu.E / e.E;
        const r21_model = ta.E / mu.E;
        const r20_model = ta.E / e.E;

        tripleContainer.innerHTML = `
            <h3>Identified lepton triple</h3>
            <table class="lepton-triple-table">
                <thead>
                    <tr>
                        <th>Family</th>
                        <th>F3 mode</th>
                        <th>Morse</th>
                        <th>E (MFT units)</th>
                        <th>Predicted (MeV)</th>
                        <th>Observed (MeV)</th>
                        <th>Error</th>
                        <th>ω²</th>
                        <th>regime</th>
                    </tr>
                </thead>
                <tbody>
                    <tr class="lepton-row electron">
                        <td><span class="dot" style="background:${LEPTON_COLORS.electron}"></span> electron</td>
                        <td>n=${e.f3_mode}</td>
                        <td>${e.morse_index} <span class="stab-tag stab-${e.stability}">${e.stability}</span></td>
                        <td>${e.E.toFixed(5)}</td>
                        <td>${e.mass_MeV.toFixed(2)}</td>
                        <td>${LEPTON_OBSERVED_MEV.electron}</td>
                        <td>calibration</td>
                        <td>${e.omega2.toFixed(4)}</td>
                        <td>${e.regime}</td>
                    </tr>
                    <tr class="lepton-row muon">
                        <td><span class="dot" style="background:${LEPTON_COLORS.muon}"></span> muon</td>
                        <td>n=${mu.f3_mode}</td>
                        <td>${mu.morse_index} <span class="stab-tag stab-${mu.stability}">${mu.stability}</span></td>
                        <td>${mu.E.toFixed(5)}</td>
                        <td>${mu.mass_MeV.toFixed(2)}</td>
                        <td>${LEPTON_OBSERVED_MEV.muon}</td>
                        <td>${errorPct(mu.mass_MeV, LEPTON_OBSERVED_MEV.muon)}%</td>
                        <td>${mu.omega2.toFixed(4)}</td>
                        <td>${mu.regime}</td>
                    </tr>
                    <tr class="lepton-row tau">
                        <td><span class="dot" style="background:${LEPTON_COLORS.tau}"></span> tau</td>
                        <td>n=${ta.f3_mode}</td>
                        <td>${ta.morse_index} <span class="stab-tag stab-${ta.stability}">${ta.stability}</span></td>
                        <td>${ta.E.toFixed(5)}</td>
                        <td>${ta.mass_MeV.toFixed(2)}</td>
                        <td>${LEPTON_OBSERVED_MEV.tau}</td>
                        <td>${errorPct(ta.mass_MeV, LEPTON_OBSERVED_MEV.tau)}%</td>
                        <td>${ta.omega2.toFixed(4)}</td>
                        <td>${ta.regime}</td>
                    </tr>
                </tbody>
            </table>
            <table class="lepton-ratios-table">
                <thead>
                    <tr><th>Mass ratio</th><th>MFT model</th><th>Observed</th><th>Error</th></tr>
                </thead>
                <tbody>
                    <tr><td>m_μ / m_e</td><td>${r10_model.toFixed(2)}</td><td>206.77</td><td>${errorPct(r10_model, 206.768)}%</td></tr>
                    <tr><td>m_τ / m_μ</td><td>${r21_model.toFixed(3)}</td><td>16.817</td><td>${errorPct(r21_model, 16.817)}%</td></tr>
                    <tr><td>m_τ / m_e</td><td>${r20_model.toFixed(1)}</td><td>3477.2</td><td>${errorPct(r20_model, 3477.2)}%</td></tr>
                </tbody>
            </table>
        `;
    } else {
        const sectorName = result.sector || 'this sector';
        const isLeptonSector = result.sector === 'lepton_sector';
        if (isLeptonSector) {
            tripleContainer.innerHTML = '<p class="hint">No three-soliton triple matching observed lepton ratios was found in the computed spectrum.</p>';
        } else {
            tripleContainer.innerHTML = `
                <h3>Spectrum for ${sectorName.replace('_', ' ')}</h3>
                <p class="hint">
                    The discrete soliton tower is shown below. Specific particle identification
                    (matching to observed quark or boson masses) is currently implemented only for
                    the charged lepton sector. The spectrum here uses Z = ${result.params.Z} from
                    the loaded preset; the same potential parameters as all other sectors.
                </p>
            `;
        }
    }

    // === Potential plot ===
    const potCanvas = document.getElementById('potential-canvas');
    const pc = result.potential_curve;
    const markers = [
        { x: result.phi_barrier, color: '#e67e22', label: `φ_b=${result.phi_barrier.toFixed(3)}` },
        { x: result.phi_vacuum, color: '#7f8c8d', label: `φ_v=${result.phi_vacuum.toFixed(3)}` },
    ];
    if (triple) {
        const e = result.spectrum[triple.electron_idx];
        const mu = result.spectrum[triple.muon_idx];
        const ta = result.spectrum[triple.tau_idx];
        markers.push({ x: e.phi_core, color: LEPTON_COLORS.electron, label: 'e⁻' });
        markers.push({ x: mu.phi_core, color: LEPTON_COLORS.muon, label: 'μ' });
        markers.push({ x: ta.phi_core, color: LEPTON_COLORS.tau, label: 'τ' });
    }
    plotCurves(potCanvas, [{ xs: pc.phi, ys: pc.V, color: '#333' }], {
        xlabel: 'φ', ylabel: 'V(φ)', markers,
    });

    // === Profile plot ===
    const profCanvas = document.getElementById('profile-canvas');
    const profHeading = document.getElementById('profile-heading');
    if (triple) {
        profHeading.textContent = 'Lepton soliton profiles u(r)';
        const e = result.spectrum[triple.electron_idx];
        const mu = result.spectrum[triple.muon_idx];
        const ta = result.spectrum[triple.tau_idx];
        // Normalize each profile for visual comparison
        const normalize = (u, r) => {
            const norm = Math.sqrt(u.reduce((sum, ui, i) => sum + ui * ui * (i > 0 ? r[i] - r[i - 1] : 0), 0));
            return norm > 0 ? u.map(ui => ui / norm) : u;
        };
        plotCurves(profCanvas, [
            { xs: e.r, ys: normalize(e.u, e.r), color: LEPTON_COLORS.electron, label: 'electron' },
            { xs: mu.r, ys: normalize(mu.u, mu.r), color: LEPTON_COLORS.muon, label: 'muon' },
            { xs: ta.r, ys: normalize(ta.u, ta.r), color: LEPTON_COLORS.tau, label: 'tau' },
        ], { xlabel: 'r', ylabel: 'u(r) / norm', xmax: 12 });
    } else {
        profHeading.textContent = 'Soliton profiles u(r)';
        clearCanvas(profCanvas);
        const ctx = profCanvas.getContext('2d');
        ctx.fillStyle = '#888';
        ctx.font = '13px sans-serif';
        ctx.textAlign = 'center';
        const sectorLabel = (result.sector || 'this sector').replace('_', ' ');
        ctx.fillText('Particle profile overlay not available for', profCanvas.width / 2, profCanvas.height / 2 - 10);
        ctx.fillText(sectorLabel + ' (yet)', profCanvas.width / 2, profCanvas.height / 2 + 10);
    }

    // === Full spectrum table ===
    const rows = result.spectrum.map((s, idx) => {
        const isLepton = s.lepton ? `lepton-row ${s.lepton}` : '';
        const leptonLabel = s.lepton ? `<span class="lepton-tag" style="background:${LEPTON_COLORS[s.lepton]}">${s.lepton}</span>` : '';
        const morseCell = (s.morse_index !== null && s.morse_index !== undefined)
            ? `${s.morse_index} <span class="stab-tag stab-${s.stability}">${s.stability}</span>`
            : '<span class="muted">—</span>';
        const f3Cell = (s.f3_mode !== null && s.f3_mode !== undefined) ? `n=${s.f3_mode}` : '';
        return `
            <tr class="${isLepton}">
                <td>${idx}${leptonLabel}</td>
                <td>${s.E.toFixed(5)}</td>
                <td>${s.mass_MeV.toFixed(2)}</td>
                <td>${s.omega2.toFixed(4)}</td>
                <td>${s.Q.toFixed(4)}</td>
                <td>${s.phi_core.toFixed(4)}</td>
                <td>${s.n_nodes}</td>
                <td>${f3Cell}</td>
                <td>${morseCell}</td>
                <td>${s.regime}</td>
            </tr>
        `;
    }).join('');

    tableContainer.innerHTML = `
        <h3>Full discrete spectrum (${result.spectrum.length} solitons, sorted by energy)</h3>
        <p class="hint">
            ω² is an eigenvalue of each soliton, not a free parameter.
            ${triple
                ? 'The canonical lepton triple is highlighted; F3 mode and Morse index columns show the Family-of-Three Stability Theorem classification (e, μ stable; τ metastable; higher modes excluded as multiply unstable).'
                : 'Specific particle identification is not implemented for this sector yet — the full discrete tower is shown.'
            }
        </p>
        <div class="spectrum-table-wrapper">
            <table class="spectrum-table">
                <thead>
                    <tr>
                        <th>#</th>
                        <th>E (MFT)</th>
                        <th>Mass (MeV)</th>
                        <th>ω²</th>
                        <th>Q</th>
                        <th>φ_core</th>
                        <th>nodes</th>
                        <th>F3 mode</th>
                        <th>Morse index</th>
                        <th>regime</th>
                    </tr>
                </thead>
                <tbody>${rows}</tbody>
            </table>
        </div>
    `;
}

// === Solver call ===

async function runSolver() {
    if (!solverReady) return;

    const solveBtn = document.getElementById('solve-btn');
    solveBtn.disabled = true;
    solveBtn.textContent = 'Computing spectrum… (this takes ~25 seconds)';

    // Clear all stale UI state up front so nothing from a previous sector lingers
    document.getElementById('lepton-triple-container').innerHTML =
        '<p class="hint" style="color:#888;">Computing spectrum…</p>';
    document.getElementById('spectrum-table-container').innerHTML = '';
    clearCanvas(document.getElementById('potential-canvas'));
    clearCanvas(document.getElementById('profile-canvas'));

    // Yield to the browser briefly so the button-disable rendering happens
    await new Promise(resolve => setTimeout(resolve, 50));

    const params = {
        m2: parseFloat(document.getElementById('m2').value),
        lam4: parseFloat(document.getElementById('lam4').value),
        lam6: parseFloat(document.getElementById('lam6').value),
        Z: parseFloat(document.getElementById('Z').value),
        a: parseFloat(document.getElementById('a').value),
        sector: activeSector,
    };

    pyodide.globals.set('js_params', pyodide.toPy(params));
    let result;
    try {
        const pyResult = pyodide.runPython('solve_spectrum(js_params)');
        result = pyResult.toJs({ dict_converter: Object.fromEntries });
        pyResult.destroy();
    } catch (err) {
        result = {
            success: false,
            message: `Python error: ${err.message}`,
            spectrum: [],
            lepton_triple: null,
            potential_curve: null,
            phi_barrier: null,
            phi_vacuum: null,
            silver_ratio: null,
            mft_to_mev: null,
        };
    }

    renderResults(result);

    solveBtn.disabled = false;
    solveBtn.textContent = 'Compute spectrum';
}

// === Preset handling ===

function loadPreset(name) {
    if (!solverReady) return;

    activeSector = name;
    pyodide.globals.set('preset_name', name);
    const presetPy = pyodide.runPython('get_preset(preset_name)');
    const preset = presetPy.toJs({ dict_converter: Object.fromEntries });
    presetPy.destroy();

    document.getElementById('m2').value = preset.m2;
    document.getElementById('lam4').value = preset.lam4;
    document.getElementById('lam6').value = preset.lam6;
    document.getElementById('Z').value = preset.Z;
    document.getElementById('a').value = preset.a;

    document.getElementById('preset-description').textContent = preset.description || '';

    // Mark active preset visually
    document.querySelectorAll('.preset').forEach(btn => {
        btn.classList.toggle('active', btn.dataset.preset === name);
    });
}

// === Wiring ===

document.addEventListener('DOMContentLoaded', () => {
    initPyodide();

    document.getElementById('solve-btn').addEventListener('click', runSolver);

    document.querySelectorAll('.preset').forEach(btn => {
        btn.addEventListener('click', e => {
            const name = e.currentTarget.dataset.preset;
            loadPreset(name);
        });
    });
});

File: qball-solver/solver.py
"""
MFT Q-Ball Solver — browser-adapted version (v2)

Adapted from mft_qball_lepton_masses.py for execution under Pyodide.
Provides solve_spectrum(params) -> dict that scans omega2 and returns
the full discrete tower of soliton solutions for a given potential.

This is the right interface for MFT: ω² is an eigenvalue, not an input.
The visitor enters the physics parameters; the equation picks out all
admissible solitons.
"""

import numpy as np
from scipy.optimize import brentq

# NumPy 2.0+ renamed trapz to trapezoid; handle both.
try:
    _trap = np.trapezoid
except AttributeError:
    _trap = np.trapz

# Calibration constant: m_e = 0.511 MeV ↔ E0 ≈ 0.00427 MFT units
M_ELECTRON_MEV = 0.511
MFT_TO_MEV_CANONICAL = 119.67  # canonical calibration at m₂=1, λ₄=2, λ₆=0.5, Z=1

# Default grid (matches mft_qball_lepton_masses.py)
RMAX_DEFAULT = 20.0
N_DEFAULT = 200


def potential(phi, m2, lam4, lam6):
    """V(φ) = m2/2·φ² − lam4/4·φ⁴ + lam6/6·φ⁶"""
    return 0.5 * m2 * phi**2 - 0.25 * lam4 * phi**4 + (1.0 / 6.0) * lam6 * phi**6


def find_barrier_and_vacuum(m2, lam4, lam6):
    """Find φ_barrier (local max of V) and φ_vacuum (second local min of V)."""
    disc = lam4**2 - 4.0 * m2 * lam6
    if disc <= 0:
        return None, None
    phi_b = float(np.sqrt((lam4 - np.sqrt(disc)) / (2.0 * lam6)))
    phi_v = float(np.sqrt((lam4 + np.sqrt(disc)) / (2.0 * lam6)))
    return phi_b, phi_v


def silver_ratio_check(m2, lam4, lam6):
    """Check whether the silver-ratio condition λ₄² = 8 m₂ λ₆ is satisfied."""
    target = 8.0 * m2 * lam6
    actual = lam4**2
    rel_err = abs(actual - target) / target if target != 0 else float('inf')
    return {
        'satisfied': rel_err < 1e-3,
        'lam4_squared': actual,
        'eight_m2_lam6': target,
        'relative_error': rel_err,
    }


def shoot(A, omega2, m2, lam4, lam6, Z, a, r, h, N):
    """
    Shoot the nonlinear Q-ball soliton equation outward from r=0.
    u ~ A*r near origin (so φ_core = A at r→0).
    Returns (u_endpoint, u_array).
    """
    u = np.zeros(N)
    u[0] = 0.0
    u[1] = A * r[1]
    for i in range(1, N - 1):
        phi_i = u[i] / r[i]
        d2u = (m2 - omega2
               - lam4 * phi_i**2
               + lam6 * phi_i**4
               - Z / np.sqrt(r[i]**2 + a**2)) * u[i]
        u[i + 1] = 2.0 * u[i] - u[i - 1] + h * h * d2u
        if not np.isfinite(u[i + 1]) or abs(u[i + 1]) > 1e8:
            u[i + 1:] = 0.0
            break
    return u[-1], u


def find_solitons_at_omega2(omega2, m2, lam4, lam6, Z, a, A_max=8.0, n_pts=300,
                             rmax=RMAX_DEFAULT, n_grid=N_DEFAULT):
    """
    At fixed omega2, scan amplitude A and find all soliton solutions
    (where u_endpoint changes sign).
    """
    r = np.linspace(rmax / (n_grid * 100.0), rmax, n_grid)
    h = r[1] - r[0]

    A_vals = np.linspace(0.01, A_max, n_pts)
    u_ends = [shoot(A, omega2, m2, lam4, lam6, Z, a, r, h, n_grid)[0]
              for A in A_vals]
    results = []

    for i in range(len(A_vals) - 1):
        if u_ends[i] * u_ends[i + 1] < 0:
            try:
                A_s = brentq(
                    lambda A: shoot(A, omega2, m2, lam4, lam6, Z, a, r, h, n_grid)[0],
                    A_vals[i], A_vals[i + 1],
                    xtol=1e-8, maxiter=50
                )
                _, u = shoot(A_s, omega2, m2, lam4, lam6, Z, a, r, h, n_grid)
                Q = float(_trap(u**2, r))
                E = omega2 * Q
                nc = int(np.sum(np.diff(np.sign(u[:int(0.95 * n_grid)])) != 0))
                phi_core = float(u[1] / r[1])
                results.append({
                    'E': float(E),
                    'Q': Q,
                    'omega2': float(omega2),
                    'A': float(A_s),
                    'n_nodes': nc,
                    'phi_core': phi_core,
                    'u': u.tolist(),
                    'r': r.tolist(),
                })
            except Exception:
                pass

    return results


def regime_for(phi_core, phi_b):
    """Classify a soliton by where its core sits relative to the barrier."""
    if phi_b is None or phi_b == 0:
        return "—"
    if phi_core < 0.5 * phi_b:
        return "linear vacuum"
    elif phi_core < 1.1 * phi_b:
        return "near-barrier"
    else:
        return "nonlinear vacuum"


def find_lepton_triple(all_solitons, R10_target=206.768, R21_target=16.817):
    """
    Among all found solitons, find the triple (E0, E1, E2) whose
    energy ratios best match the observed lepton mass ratios.
    Replicates the best_triple logic from mft_qball_lepton_masses.py.

    Returns: dict with keys electron, muon, tau (each a soliton dict
             from the spectrum) plus 'score' indicating fit quality,
             or None if no good triple found.
    """
    if len(all_solitons) < 3:
        return None

    best_score = float('inf')
    best_triple = None

    for i in range(len(all_solitons)):
        for j in range(i + 1, len(all_solitons)):
            for k in range(j + 1, len(all_solitons)):
                E0 = all_solitons[i]['E']
                E1 = all_solitons[j]['E']
                E2 = all_solitons[k]['E']
                if E0 > 0:
                    score = (np.log(E1 / E0 / R10_target))**2 + \
                            (np.log(E2 / E1 / R21_target))**2
                    if score < best_score:
                        best_score = score
                        best_triple = (i, j, k)

    if best_triple is None:
        return None

    i, j, k = best_triple
    return {
        'electron_idx': i,
        'muon_idx': j,
        'tau_idx': k,
        'score': float(best_score),
    }


def solve_spectrum(params):
    """
    Main entry point. Scan ω² values and return the full discrete tower
    of soliton solutions for the given potential.

    This is the canonical mode: visitor enters physics parameters,
    receives the discrete spectrum back as a single result.

    params: dict with keys m2, lam4, lam6, Z, a (and optionally
            n_omega for scan resolution, default 40)

    Returns: dict with keys
        success: bool
        message: str
        spectrum: list of soliton dicts, sorted by energy, deduplicated
        phi_barrier, phi_vacuum: potential landscape
        potential_curve: {phi: [...], V: [...]}
        silver_ratio: {satisfied, lam4_squared, eight_m2_lam6, relative_error}
        params: the inputs (echoed back)
    """
    try:
        m2 = float(params.get('m2', 1.0))
        lam4 = float(params.get('lam4', 2.0))
        lam6 = float(params.get('lam6', 0.5))
        Z = float(params.get('Z', 1.0))
        a = float(params.get('a', 1.0))
        n_omega = int(params.get('n_omega', 40))

        # Potential landscape
        phi_b, phi_v = find_barrier_and_vacuum(m2, lam4, lam6)

        # Sample V(φ) for plotting
        phi_max = (phi_v * 1.4) if phi_v else 3.0
        phi_arr = np.linspace(0, phi_max, 200)
        V_arr = potential(phi_arr, m2, lam4, lam6)

        # Silver-ratio diagnostic
        sr = silver_ratio_check(m2, lam4, lam6)

        # Scan omega2 and collect all soliton solutions
        all_solitons = []
        for omega2 in np.linspace(0.05, 0.99, n_omega):
            sols = find_solitons_at_omega2(omega2, m2, lam4, lam6, Z, a)
            for s in sols:
                # Deduplicate by energy (matches original mft_qball_lepton_masses.py logic)
                if not any(abs(s['E'] - prev['E']) < 0.01 for prev in all_solitons):
                    all_solitons.append(s)

        # Sort by energy
        all_solitons.sort(key=lambda x: x['E'])

        # Annotate with regime
        for s in all_solitons:
            s['regime'] = regime_for(s['phi_core'], phi_b)

        # Compute mass calibration: assume the lowest-energy soliton is the electron
        # (calibration anchor). All other masses scale relative to it.
        if all_solitons and all_solitons[0]['E'] > 0:
            E0 = all_solitons[0]['E']
            scale = M_ELECTRON_MEV / E0  # MeV per MFT energy unit
            for s in all_solitons:
                s['mass_MeV'] = s['E'] * scale
        else:
            scale = MFT_TO_MEV_CANONICAL
            for s in all_solitons:
                s['mass_MeV'] = s['E'] * scale

        # Identify the canonical lepton triple (e, μ, τ) by best ratio match
        triple = find_lepton_triple(all_solitons)
        if triple is not None:
            # Tag the chosen solitons
            all_solitons[triple['electron_idx']]['lepton'] = 'electron'
            all_solitons[triple['muon_idx']]['lepton'] = 'muon'
            all_solitons[triple['tau_idx']]['lepton'] = 'tau'

        return {
            'success': True,
            'message': f"Found {len(all_solitons)} distinct solitons in the spectrum.",
            'spectrum': all_solitons,
            'lepton_triple': triple,
            'phi_barrier': phi_b,
            'phi_vacuum': phi_v,
            'potential_curve': {
                'phi': phi_arr.tolist(),
                'V': V_arr.tolist(),
            },
            'silver_ratio': sr,
            'mft_to_mev': float(scale),
            'params': {
                'm2': m2, 'lam4': lam4, 'lam6': lam6,
                'Z': Z, 'a': a,
            },
        }
    except Exception as e:
        return {
            'success': False,
            'message': f"Error: {type(e).__name__}: {e}",
            'spectrum': [],
            'phi_barrier': None,
            'phi_vacuum': None,
            'potential_curve': None,
            'silver_ratio': None,
            'mft_to_mev': None,
            'params': params,
        }


# Canonical preset: lepton sector with silver-ratio condition
PRESETS = {
    'lepton_sector': {
        'm2': 1.0,
        'lam4': 2.0,
        'lam6': 0.5,
        'Z': 1.0,
        'a': 1.0,
        'description': (
            'Lepton sector: silver-ratio potential (m₂=1, λ₄=2, λ₆=0.5 satisfies '
            'λ₄² = 8m₂λ₆), Z=1 Coulomb coupling, a=1 softening length. '
            'The Q-ball equation produces the full charged-lepton spectrum '
            '(e, μ, τ) from this single parameter set. Energy scale calibrated '
            'by setting the lowest soliton energy equal to m_e = 0.511 MeV.'
        ),
    },
}


def get_preset(name):
    """Return preset parameters by name."""
    return PRESETS.get(name, PRESETS['lepton_sector'])
